virsh_destroy: Add testcase for BZ#2045879

liuyd96 opened the pull request:
Signed-off-by: Liu Yiding <liuyd.fnst@fujitsu.com>


libvirt-8.0.0-2
```
(1/1) type_specific.io-github-autotest-libvirt.virsh.destroy.normal_test.non_acl.no_option.clean_domain_socket: FAIL: Failed to start guest: error: Failed to start domain 'avocado-vt-vm1'\nerror: cannot open /dev/null: Too many open files\n (72.72 s)
```

libvirt-8.0.0-5
```
 (1/1) type_specific.io-github-autotest-libvirt.virsh.destroy.normal_test.non_acl.no_option.clean_domain_socket: PASS (122.57 s)
```

liuyd96 commented:
Hi, @iccaszhulili @Yingshun 
Please help review it. Thanks.

iccaszhulili commented:
> Hi, @iccaszhulili @Yingshun Please help review it. Thanks.

Hi, Yiding

From the test result, I can not tell whether there are test result on RHEL9 OS

iccaszhulili commented:
> Hi Lili
> 
> > > Hi, @iccaszhulili @Yingshun Please help review it. Thanks.
> > 
> > 
> > Hi, Yiding
> > From the test result, I can not tell whether there are test result on RHEL9 OS
> 
> The test result was from RHEL8 OS kernel-4.18.0-369.el8.aarch64. Thanks.

Could you please provide the testing results on RHEL9 OS. Since on RHEL8,  we will config libvirtd.service file; while on RHEL9, we will config virtqemud.service. It's better to give testing results on RHEL9 either. Thanks

chunfuwen commented on libvirt/tests/src/virsh_cmd/domain/virsh_destroy.py:
```diff
@@ -10,6 +16,11 @@
 
+# Using as lower capital is not the best way to do, but this is just a
+# workaround to avoid changing the entire file.
+logging = log.getLogger('avocado.' + __name__)
```
since it is workaround, so it is time to change it by following reference below:
```
import logging
LOG = logging.getLogger('avocado.' + __name__)
LOG.debug(xx)
```

liuyd96 replied:
> since it is workaround, so it is time to change it by following reference below:
> 
> ```
> import logging
> LOG = logging.getLogger('avocado.' + __name__)
> LOG.debug(xx)
> ```

Done

chunfuwen commented on libvirt/tests/src/virsh_cmd/domain/virsh_destroy.py:
```diff
@@ -51,6 +64,23 @@ def run(test, params, env):
+        :param modify_info: libvirtd modify info
+        """
+        ori_value = process.getoutput("cat %s | grep LimitNOFILE" % libvirtd_file, shell=True)
+        with open(libvirtd_file, 'r') as fr:
```
It sound like we can combine read or write in one with ... open statement.
One reference could be https://stackoverflow.com/questions/2424000/read-and-overwrite-a-file-in-python

liuyd96 replied:
> It sound like we can combine read or write in one with ... open statement. One reference could be https://stackoverflow.com/questions/2424000/read-and-overwrite-a-file-in-python

Got it. Thanks.

Yingshun commented on libvirt/tests/src/virsh_cmd/domain/virsh_destroy.py:
```diff
@@ -74,6 +104,33 @@ def run(test, params, env):
+            status = 0
+            try:
+                # Backup original libvirtd.service
+                libvirtd_file = "/usr/lib/systemd/system/libvirtd.service"
```
Do you need to modify virtqemud.service when modular daemons are enabled? @iccaszhulili Could you please double check?

iccaszhulili replied:
Yes, should config /usr/lib/systemd/system/virtqemud.service on RHEL9

liuyd96 replied:
> Yes, should config /usr/lib/systemd/system/virtqemud.service on RHEL9

Done. BTW, RHEL9 libvirt-8.0.0-2 doesn't have this BUG since the default glib2 has fixed this error.

liuyd96 commented on libvirt/tests/src/virsh_cmd/domain/virsh_destroy.py:
```diff
@@ -1,7 +1,13 @@
 #import logging
```
> You're going to import logging, so would you please remove this comment?

Done

liuyd96 commented:
CI check have failed due to spell-check  of "virtqemud". Ignore it.

- On RHEL8 OS

libvirt-8.0.0-2.el8.aarch64
```
(1/1) type_specific.io-github-autotest-libvirt.virsh.destroy.normal_test.non_acl.no_option.clean_domain_socket: FAIL: Failed to start guest: error: Failed to start domain 'avocado-vt-vm1'\nerror: cannot open /dev/null: Too many open files\n (73.19 s)
```
libvirt-8.0.0-5.module+el8.6.0+14344+04da0821.aarch64
```
(1/1) type_specific.io-github-autotest-libvirt.virsh.destroy.normal_test.non_acl.no_option.clean_domain_socket: PASS (120.05 s)
```

- On RHEL9 OS. 

libvirt-8.0.0-5.el9.aarch64
```
(1/1) type_specific.io-github-autotest-libvirt.virsh.destroy.normal_test.non_acl.no_option.clean_domain_socket: PASS (119.08 s)
```

libvirt-8.0.0-2.el9.aarch64. The RHEL9 default glib2 has fix this bug so  test passed on libvirt-8.0.0-2.
```
(1/1) type_specific.io-github-autotest-libvirt.virsh.destroy.normal_test.non_acl.no_option.clean_domain_socket: PASS (121.08 s)
```

liuyd96 commented:
Hi @iccaszhulili.
The 're-request review' option doesn't work with you. Please help check it if you are free. Thanks.

chunfuwen commented on libvirt/tests/src/virsh_cmd/domain/virsh_destroy.py:
```diff
@@ -74,6 +100,49 @@ def run(test, params, env):
+                if distro.detect().name == 'rhel' and \
+                        int(distro.detect().version) < 9:
+                    modify_service = "libvirtd.service"
+                service_file = "/usr/lib/systemd/system/{}"\
```
Perhaps, we can get service file by command dynamically, rather than hard code string:
[root@hp-dl360g9-16 ~]# systemctl show -p FragmentPath libvirtd.service
FragmentPath=/usr/lib/systemd/system/libvirtd.service

Yingshun commented on libvirt/tests/src/virsh_cmd/domain/virsh_destroy.py:
```diff
@@ -74,6 +100,49 @@ def run(test, params, env):
+                # Decrease domain number to speed up test
+                modify_virtual_daemon(service_file, limit_nofile)
+                process.run("systemctl daemon-reload")
+                if "virtqemud.service" in modify_service:
+                    utils_split_daemons.virtqemud_restart()
+                else:
+                    utils_libvirtd.libvirtd_restart()
```
I would recommend using utils_libvirtd.Libvirtd() directly.
```suggestion
              utils_libvirtd.Libvirtd('virtqemud').restart()  
```

Yingshun commented on libvirt/tests/src/virsh_cmd/domain/virsh_destroy.py:
```diff
@@ -74,6 +100,49 @@ def run(test, params, env):
+                # Recover libvirtd.service
+                shutil.copy(backup_file, service_file)
+                process.run("systemctl daemon-reload")
+                if "virtqemud.service" in modify_service:
+                    utils_split_daemons.virtqemud_restart()
+                else:
+                    utils_libvirtd.libvirtd_restart()
```
Same as above.

Yingshun commented on libvirt/tests/src/virsh_cmd/domain/virsh_destroy.py:
```diff
@@ -74,6 +100,49 @@ def run(test, params, env):
+        if start_destroy_times:
+            status = 0
+            try:
+                # RHEL9 will config virtqemud.service
+                modify_service = "virtqemud.service"
+                if distro.detect().name == 'rhel' and \
+                        int(distro.detect().version) < 9:
+                    modify_service = "libvirtd.service"
```
We should get configuration file through libvirt daemons mode, but not distribution version  as I know. So I think it's better to update it to
```suggestion
                modify_service = utils_libvirtd.Libvirtd('virtqemud').service_name
```
OR 
```suggestion
                modify_service = "virtqemud.service"
                if utils_split_daemons.is_modular_daemon():
                      modify_service = "libvirtd.service"
```
What do you think?

iccaszhulili commented on libvirt/tests/src/virsh_cmd/domain/virsh_destroy.py:
```diff
@@ -74,6 +100,49 @@ def run(test, params, env):
+                # RHEL9 will config virtqemud.service
+                modify_service = "virtqemud.service"
+                if distro.detect().name == 'rhel' and \
+                        int(distro.detect().version) < 9:
```
Hi, 

We need to config virtqemud.service under split deamon mode. We can use utils_split_daemons.is_modular_daemon() to tell whether the current daemon mode is split daemon or monolithic daemon.

liuyd96 commented:
- RHEL8 + libvirt-8.0.0-5
```
(1/1) type_specific.io-github-autotest-libvirt.virsh.destroy.normal_test.non_acl.no_option.clean_domain_socket: PASS (118.56 s)
```

- RHEL8 + libvirt-8.0.0-2
```
(1/1) type_specific.io-github-autotest-libvirt.virsh.destroy.normal_test.non_acl.no_option.clean_domain_socket: FAIL: Failed to start guest: error: Failed to start domain 'avocado-vt-vm1'\nerror: cannot open /dev/null: Too many open files\n (81.33 s)
```
- RHEL9 + libvirt-8.0.0-5
```
(1/1) type_specific.io-github-autotest-libvirt.virsh.destroy.normal_test.non_acl.no_option.clean_domain_socket: PASS (124.74 s)
```

Yingshun reviewed (approved):
LGTM

iccaszhulili commented:
LGTM

liuyd96 commented:
@chunfuwen gently ping~

chunfuwen reviewed (approved):
LGTM

liuyd96 commented:
> One comment not addressed, but not block approved LGTM

Oops，my response was in Pending Status and I didn't submit it... 
Anyway, I think I have addressed all comments. Thanks.
